### ISSUE TITLE
Problems with forward compatibility

### ISSUE THREAD

**Issue by voicesmackiev** [aws-amplify/amplify-swift#2865]
### Describe the bug

Our app uses DataStore to save/query models from storage. Properties which were added to newer versions of app are reset to nil if models are modified by app with older versions. Expected result - models should be merged to save properties, not supported by previous versions. How this can be fixed?

### Steps To Reproduce

```swift
Assume app with version 1 works with 
public struct CustomModel: Model {
  public let id: String
  public var property1: String?
}

Later in version 2 the app adds new optional property2 to the model:
public struct CustomModel: Model {
  public let id: String
  public var property1: String?
  public var property2: String?
}

1. App with version 2 saves CustomModel where property2 = "some string"
2. App with version 1 queries this record, makes some changes and saves it to the storage
3. App with version 2 queries this record and try to read property2

Actual behavior:
property2 is nil
```

### Expected behavior

property2 should be "some string", it should not be reset by older versions

### Amplify Framework Version

1.28.1

### Amplify Categories

API, Auth, DataStore, Storage

### Dependency manager

Swift PM

### Swift version

5.0

### CLI version

10.7.3

### Xcode version

14.3 (14E222b)

### Relevant log output

```shell
<details>
<summary>Log Messages</summary>


INSERT LOG MESSAGES HERE
```
</details>
```


### Is this a regression?

No

### Regression additional context

_No response_

### Device

different iPads and iPhones

### iOS Version

different iOS versions

### Specific to simulators

No

### Additional context

_No response_

**Comment by voicesmackiev** [comment]
Hi @jcjimenez! Can you please advise: is it possible to workaround this issue? On app or on server side? We already released apps with Amplify support and forward compatibility is important for our functionality.

**Comment by lawmicha** [comment]
Hi @voicesmackiev, I couldn't reproduce this, but here's a few thoughts to continue debugging. Let us know if you have more info.

> 1. App with version 2 saves CustomModel where property2 = "some string"

After this step, can you verify from another client like the AppSync Console and query that the model has the property2 string?

> 2. App with version 1 queries this record, makes some changes and saves it to the storage

At this point, if you have verbose logging enabled (`Amplify.Logging.logLevel = .verbose`), you should be able to see the "sending mutation" sync the data from local storage to AppSync using an UpdateMutation. The update mutation should have a variables object containing all the fields from the model. Since it's v1, the property2 key-value doesn't exist: `variables: Optional(["input": ["_version": Optional(1), "property1": Optional(Optional("updated property 1 from v1")), "id": Optional("111")]])`

```swift
sending mutation with sync data: GraphQLRequest<MutationSync<AnyModel>>(
  apiName: nil, 
  document: "mutation UpdateCustomModel($input: UpdateCustomModelInput!) {\n  updateCustomModel(input: $input) {\n    id\n    createdAt\n    property1\n    updatedAt\n    __typename\n    _version\n    _deleted\n    _lastChangedAt\n  }\n}", 
  variables: Optional(["input": ["_version": Optional(1), "property1": Optional(Optional("updated property 1 from v1")), "id": Optional("111")]]), responseType: AWSPluginsCore.MutationSync<AWSPluginsCore.AnyModel>, 
  decodePath: Optional("updateCustomModel"), 
  options: Optional(Amplify.GraphQLRequest<AWSPluginsCore.MutationSync<AWSPluginsCore.AnyModel>>.Options(pluginOptions: Optional(AWSPluginsCore.AWSPluginOptions(authType: nil, modelName: Optional("CustomModel"))))))
```
Since `property2` doesn't get set to `nil` from v1 app client, this property should remain unchanged.

> 3. App with version 2 queries this record and try to read property2

Unfortunately I wasn't able to reproduce the issue, in my sample app v2 I was able to query an instance of CustomModel, even after clearing the local database with `DataStore.clear()` and in AppSync Console I can query the model and property2 is unchanged.

**Comment by voicesmackiev** [comment]
Hi @lawmicha , sorry for late response, please see attached schema and configuration files:
[Archive.zip](https://github.com/aws-amplify/amplify-swift/files/12323644/Archive.zip)

- The issue is reproducible with Amplify.DataStore.query/save only, checked with the latest 1.30.4
- new properties are NOT nil if check the instance of Coach at DynamoDB > Explore items, they are valid where
- Local xxx.db in Documents has nil values after #6
- The issue is NOT reproducible if use Amplify.API to query data.

1. App with old version of schema (schema_old.graphql) is installed on device A
2. App with new version of schema (schema_nwe.graphql) is installed on device B
3. Create instance of Coach on old A
4. Fetch the Coach on new B and modify new 'propertyStr' or 'propertyBool'
5. Fetch the Coach on old A; modify old 'name' property and save Amplify.DataStore.save
6. Fetch the Coach on new A and observe values in new 'propertyStr' or 'propertyBool' - they are nil. They are also nil in loca .db file.

**Comment by lawmicha** [comment]
Thanks for the details, there are a lot of bug fixes and active development on v2. Would you be able to upgrade to 2.16.0? https://github.com/aws-amplify/amplify-swift/releases/tag/2.16.0

I've been thinking through the steps you described, and I believe I've identified the issue. Even updating to v2 will not fix this problem.

## Problem

Upgrading a schema with optional values may still have undesirable behaviors when persisting data from the device with the previous schema.

Device A - old schema
Device B - new schema

Device B establishes subscriptions to the Coach model when DataStore is started. All subscription events to B, sourced from device A, will be missing the values for the new fields. Device A inadvertedly update it's local copy of the data of the new fields to nil, and subsequent updates to persist the nil values.

### Notes

Device B subscription selection set
```
{  
   id
  name
  owner
  propertyStr
  propertyBool
}
```
Device A mutation's selection set
```
{
   id
  name
  owner
}
```

Device A's mutation response
```
{
 "data": {
    "id": "coach123",
     "name": "newName",
     "owner: "owner",
  }
}
```

Data in the DynamoDB
```
| id | name | owner| propertyStr | propertyBool
| "coach123" | "newName" | "owner" | "somePropertyStr" | true
```

Because the mutation response for Device's A's mutation is missing the new fields, Device B thinks they are nil and updates the local database with `nil` values. Any subsequent changes to this data model will send the nil values of `propertyStr` and `propertyBool` and update the database.

```
| id | name | owner| propertyStr | propertyBool
| "coach123" | "newName" | "owner" | nil | nil
```

**Comment by lawmicha** [comment]
Cross referencing to doc updates: https://github.com/aws-amplify/docs/issues/5873 

@voicesmackiev since we have identified the root cause of this issue, i'll close this and continue to draft up the content over in https://github.com/aws-amplify/docs/issues/5873 feel free to open another issue if you are facing something else or have more questions
